torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp

```diff
@@ -24,36 +24,49 @@ using namespace ::torch::jit;
 // conversion module into a group. We should re-implement this.
 std::vector<Node*> FakeCluster(const std::shared_ptr<Graph>& graph) {
   std::vector<Node*> nodes;
-  for (auto node : graph->nodes()) {
+  auto is_disc_compilable = [](torch::jit::Node* node) {
     if (torch::blade::IsMlirMhloSupported(*node) &&
         node->kind() != prim::Constant) {
-      if (node->kind() == aten::addmm) continue;
-      nodes.push_back(node);
+      for (auto& input : node->inputs()) {
+        // input should be Tensor or Scalar with explict type
+        auto typ = input->type();
+        if (!typ->cast<c10::TensorType>() &&
+            c10::tryScalarTypeFromJitType(*typ) == c10::nullopt)
+          return false;
+      }
+      return true;
     }
-  }
+    return false;
+  };
+
+  std::copy_if(graph->nodes().begin(), graph->nodes().end(),
+               std::back_inserter(nodes), is_disc_compilable);
   return nodes;
 }
 
 // Give:
 //  with prim::FusionGroup(
-//      %1: Scalar,
-//      %2: Scalar):
-//    %3 Tensor = aten::add(%1, %2)
-//  return %3
+//      %p0: Tensor,
+//      %p1: Tensor,
+//      %p2: Scalar):
+//    %1 Tensor = aten::add(%p0, %p1, %p2)
+//  return %1
 //
 // Execute: CastToTensorInputs(sub_graph)
 //
 // After:
 //  with prim::FusionGroup(
-//      %1.1: Tensor,
-//      %2.1: Tensor):
-//    %4 = aten::item(%1.1, 1)
-//    %5 = aten::item(%2.1, 1)
-//    %3 Tensor = aten::add(%4, %5)
-//    return %3
-void CastBoundaryScalarToTensor(Graph* disc_graph, size_t i) {
+//      %p0.1: Tensor,
+//      %p1.1: Tensor,
+//      %p2.1: Tensor):
+//    %1 int = aten::item(%p2.1)
+//    %2 Tensor = aten::add(%p0.1, %p1.1, %1)
+//    return %2
+void CastBoundaryScalarToTensor(Graph* disc_graph, size_t i,
+                                at::ScalarType& typ) {
   auto new_input = disc_graph->insertInput(
       i, c10::string(disc_graph->inputs()[i]->debugName() + ".1"));
+  new_input->setType(TensorType::create(typ, c10::nullopt, 0, false));
```

Comment:
Should `item_node->output()` be set to the same type? (aka line 73)

Reply:
Done. Thanks, we should do that.